### PR TITLE
Remove support for old translog checkpoint formats

### LINKED ISSUE #44210: Remove support for old translog checkpoint formats
Today we support three formats of checkpoint. We stopped using the oldest format in 6.0, so support for this can be dropped in both `7.x` and `master`. We stopped using the second-oldest format in 6.4, so support for this can be dropped in `master`. I also don't see any tests for Elasticsearch's behaviour with checkpoint files of these old formats (like e.g. `TranslogHeaderTests#testLegacyTranslogVersions`), but I think we should have some.

### PR DESCRIPTION
**Pull request by Hohol**
Remove support for old translog checkpoint formats.
Add unit test for Elasticsearch's behaviour with checkpoint files of old formats.

Closes https://github.com/elastic/elasticsearch/issues/44210

### CODE REVIEW

**Comment by Hohol** [comment]
@DaveCTurner, am I on the right track? Is there anything else to be done in master (besides tests)?

**Comment by DaveCTurner** [comment]
Yes that's about all that's needed for the production code in `master`, except that a few of the comments still mention version numbers or otherwise refer back to older versions. It's all about the tests :)

File: server/src/main/java/org/elasticsearch/index/translog/Checkpoint.java [hunk 1]
```diff
@@ -188,18 +143,10 @@ public static Checkpoint read(Path path) throws IOException {
...
-                    return Checkpoint.readCheckpointV6_0_0(indexInput);
-                } else {
-                    assert fileVersion == CURRENT_VERSION : fileVersion;
-                    assert indexInput.length() == V3_FILE_SIZE : indexInput.length();
-                    return Checkpoint.readCheckpointV6_4_0(indexInput);
-                }
+                final int fileVersion = CodecUtil.checkHeader(indexInput, CHECKPOINT_CODEC, CURRENT_VERSION, CURRENT_VERSION);
+                assert fileVersion == CURRENT_VERSION : fileVersion;
+                assert indexInput.length() == V3_FILE_SIZE : indexInput.length();
+                return Checkpoint.readCheckpointV6_4_0(indexInput);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/translog/Checkpoint.java]: Suggest removing the mention of a specific (ancient) version here too:

```suggestion
                return Checkpoint.readCheckpointV3(indexInput);
```

File: server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 2]
```diff
@@ -2785,6 +2786,13 @@ public void testCheckpointOnDiskFull() throws IOException {
         assertEquals(read, checkpoint);
     }
 
+    public void testLegacyCheckpointVersion() {
+        expectThrows(
+            IndexFormatTooOldException.class,
+            () -> Checkpoint.read(getDataPath("/org/elasticsearch/index/checkpoint/v2.cpk"))
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: :+1: although #44217 means this now throws a different exception.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Thanks @Hohol looks good, but needs merging with latest master.

**Comment by DaveCTurner** [comment]
@elasticmachine test this please

File: server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 3]
```diff
@@ -2791,6 +2792,14 @@ public void testCheckpointOnDiskFull() throws IOException {
         assertEquals(read, checkpoint);
     }
 
+    public void testLegacyCheckpointVersion() {
+        expectThrows(
+            TranslogCorruptedException.class,
+            IndexFormatTooOldException.class,
+            () -> Checkpoint.read(getDataPath("/org/elasticsearch/index/checkpoint/v2.cpk"))
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: Filename typo:

```suggestion
            () -> Checkpoint.read(getDataPath("/org/elasticsearch/index/checkpoint/v2.ckp"))
```

**Comment by Hohol** [comment]
Forgot to actually change the file name...

**Comment by DaveCTurner** [comment]
Oh hang on the build will still fail after renaming that file - try the `:server:forbiddenpatterns` task. Stand by I will work out what pattern would be acceptable.

**Comment by DaveCTurner** [comment]
Please call it `v2.ckp.binary`.

**Comment by DaveCTurner** [comment]
@elasticmachine please test this

**Comment by DaveCTurner** [comment]
@elasticmachine test this please

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/2

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**